transport/ssh.go

```diff
@@ -39,51 +40,139 @@ func newAgent() (ssh.AuthMethod, error) {
+		var f *os.File
+		if f, err = os.OpenFile(id.KnownHosts, os.O_RDONLY|os.O_CREATE, 0644); err != nil {
+			return
+		}
+		f.Close()
+
+		conf.hostKeyCallback, err = knownhosts.New(id.KnownHosts)
+		if err != nil {
+			err = fmt.Errorf("Loading SSH known hosts file: %v", err)
+			return
+		}
+		conf.hostKeyCallback = acceptNewCallback(ctx, conf.hostKeyCallback, id.KnownHosts)
+
+		// Lookup host key alias for use in key checking. This should be the last wrapped so that
+		// other callbacks use the alias.
+		if id.HostKeyAlias == "" {
+			if id.HostKeyAlias, err = ssh_config.GetStrict(id.Host, "HostKeyAlias"); err != nil {
+				return
+			}
+		}
+		if id.HostKeyAlias != "" {
+			conf.hostKeyCallback = hostAliasCallback(conf.hostKeyCallback, id.HostKeyAlias)
+		}
+	}
+	return
 }
 
-func sshConnect(ctx context.Context, host, port, user string, identityfile string, strictHostKeyChecking bool) (*ssh.Client, error) {
-	connID := user + "@" + host + ":" + port
+func sshConnect(ctx context.Context, conf connInfo, retries uint) (*ssh.Client, error) {
+	connID := conf.user + "@" + conf.host + ":" + conf.port
 	// This is a single-use cache, so pass in an empty category.
 	obj, err := connectionCache.GetOrUpdate("", connID, expires, true, func() (interface{}, error) {
 		agent, err := newAgent()
 		if err != nil {
 			return nil, fmt.Errorf("Failed to find config from SSH_AUTH_SOCK: %v", err)
 		}
 
-		hostKeyCallback := ssh.InsecureIgnoreHostKey()
-		if strictHostKeyChecking {
-			hostKeyCallback, err = getHostKeyCallback()
-			if err != nil {
-				return nil, fmt.Errorf("Loading SSH known hosts file: %v", err)
-			}
-		}
-
 		var authmethod []ssh.AuthMethod
-		if key, err := ioutil.ReadFile(identityfile); err != nil {
-			activity.Record(ctx, "Unable to read private key, falling back to SSH agent: %v", err)
-		} else {
-			if signer, err := ssh.ParsePrivateKey(key); err != nil {
-				activity.Record(ctx, "Unable to parse private key, falling back to SSH agent: %v", err)
+		for _, identityFile := range conf.identityFiles {
+			if key, err := ioutil.ReadFile(identityFile); err != nil {
+				activity.Record(ctx, "Unable to read private key, falling back to SSH agent: %v", err)
 			} else {
-				authmethod = append(authmethod, ssh.PublicKeys(signer))
+				if signer, err := ssh.ParsePrivateKey(key); err != nil {
+					activity.Record(ctx, "Unable to parse private key, falling back to SSH agent: %v", err)
+				} else {
+					authmethod = append(authmethod, ssh.PublicKeys(signer))
+				}
 			}
 		}
 		// Append agent now so that it comes last in case we find another method to try.
 		authmethod = append(authmethod, agent)
 		sshConfig := &ssh.ClientConfig{
-			User:            user,
+			User:            conf.user,
 			Auth:            authmethod,
-			HostKeyCallback: hostKeyCallback,
+			HostKeyCallback: conf.hostKeyCallback,
+		}
+
+		// Retry until period expires or connection is established.
+		if retries > 0 {
```

Comment:
Nit: You could also do `if retries <= 0 { retries = 1 }` to centralize this code.

Reply:
I don't think that would have the same meaning.